runtime/memprof.c

```diff
@@ -334,264 +346,322 @@ struct tracked {
-        *trackst.entries[i].idx_ptr = j;
-      trackst.entries[j] = trackst.entries[i];
+  uintnat i, j;
+
+  if (ea == NULL) return;
+
+  j = i = ea->delete_idx;
+  while (i < ea->len) {
+    if (!ea->t[i].deleted) {
+      if (ea->t[i].callback_running) {
+        if (callback_status == i) callback_status = j;
+        if (caml_memprof_th_ctx_iter_hook != NULL) {
+          struct th_ctx_change_callback_status_data closure = { i, j };
+          caml_memprof_th_ctx_iter_hook(th_ctx_change_cb_st, &closure);
+        }
+      }
+      ea->t[j] = ea->t[i];
       j++;
     }
     i++;
-    if (trackst.young == i) trackst.young = j;
-    if (trackst.callback == i) trackst.callback = j;
+    if (ea->young_idx == i) ea->young_idx = j;
+    if (ea == &entries_global && callback_idx == i) callback_idx = j;
   }
-  trackst.delete = trackst.len = j;
-  CAMLassert(trackst.callback <= trackst.len);
-  CAMLassert(trackst.young <= trackst.len);
-  realloc_trackst();
+  ea->delete_idx = ea->len = j;
+  CAMLassert(ea != &entries_global || callback_idx <= ea->len);
+  CAMLassert(ea->young_idx <= ea->len);
+  realloc_entries(ea, 0);
 }
 
-static void check_action_pending(void) {
-  if (!suspended && trackst.callback < trackst.len)
+static void check_action_pending(void)
+{
+  if (suspended) return;
+  if (callback_idx < entries_global.len ||
+      (entries_local != NULL && entries_local->len > 0))
     caml_set_action_pending();
 }
 
-void caml_memprof_set_suspended(int s) {
+void caml_memprof_set_suspended(int s)
+{
   suspended = s;
   caml_memprof_renew_minor_sample();
-  if(!s) check_action_pending();
+  if (!s) check_action_pending();
 }
 
 /* In case of a thread context switch during a callback, this can be
    called in a reetrant way. */
 value caml_memprof_handle_postponed_exn(void)
 {
   value res = Val_unit;
-  if (suspended || trackst.callback >= trackst.len) return res;
+  uintnat i;
+  if (suspended) return Val_unit;
+  if (callback_idx >= entries_global.len &&
+      (entries_local == NULL || entries_local->len == 0)) return Val_unit;
 
   caml_memprof_set_suspended(1);
-  while (trackst.callback < trackst.len) {
-    uintnat i = trackst.callback;
-    res = handle_entry_callbacks_exn(&i);
-    if (Is_exception_result(res)) break;
+
+  if (entries_local != NULL)
+    for (i = 0; i < entries_local->len; i++) {
+      /* We are the only thread allowed to modify [entries_local], so
+         the indices cannot shift. */
+      res = run_alloc_callback_exn(i);
+      if (Is_exception_result(res)) goto end;
+      if (realloc_entries(&entries_global, 1))
+        /* Transfer the entry to the global array. */
+        entries_global.t[entries_global.len++] = entries_local->t[i];
+      mark_deleted(entries_local, i);
+    }
+
+  while (callback_idx < entries_global.len) {
+    struct tracked* t = &entries_global.t[callback_idx];
+
+    if (t->deleted || t->callback_running) {
+      /* This entry is not ready. Ignore it. */
+      callback_idx++;
```

Comment:
I think there's a small bug here. Suppose another thread is blocked in a promotion callback on callback_idx = 2. This code will then set callback_idx = 3. When the other thread finishes its promotion callback, it won't consider running the deallocation callback for callback_idx = 2, because the loop in caml_memprof_handle_postponed_exn will move on.

Reply:
Thanks. There is a fix in 4418ffdb. This is a bit hacky, since it involves checking for a special case in `run_callback_exn`. If you have a better idea, then I'm open, but I am afraid it involves adding quite a bit of complexity.

Reply:
Thanks. I suspect there may be a simpler way to do this, I'll see if I can come up with one. Your fix looks correct, though.